fe/fe-core/src/main/java/org/apache/iceberg/MetadataParser.java

```diff
@@ -0,0 +1,338 @@
+
+    private final ThreadLocal<Kryo> kryoThreadLocal = ThreadLocal.withInitial(() -> {
+        Kryo kryo = new Kryo();
+        kryo.register(CommonMetadataBean.class);
+        kryo.register(IcebergMetricsBean.class);
+        UnmodifiableCollectionsSerializer.registerSerializers(kryo);
+        return kryo;
+    });
+
+    public MetadataParser(Table table,
+                          Map<Integer, String> specStringCache,
+                          Map<Integer, ResidualEvaluator> residualCache,
+                          ExecutorService executorService,
+                          ScanMetrics scanMetrics,
+                          DeleteFileIndex deleteFileIndex,
+                          MetadataCollectJob job,
+                          long liveFilesCount) {
+        this.table = table;
+        this.job = job;
+        this.metrics = scanMetrics;
+        this.deleteFileIndex = deleteFileIndex;
+        this.isPartitionedTable = table.spec().isPartitioned();
+        this.specStringCache = specStringCache;
+        this.residualCache = residualCache;
+        this.schemaString = SchemaParser.toJson(table.schema());
+        this.executorService = executorService;
+        this.liveFilesCount = liveFilesCount;
+    }
+
+    public void parse() {
+        List<Future<Boolean>> futures = new ArrayList<>();
+        ConnectContext context = job.getContext();
+        Queue<TResultBatch> resultBatchQueue = job.getResultQueue();
+        while (context.getState().isRunning() || !resultBatchQueue.isEmpty()) {
+            if (!resultBatchQueue.isEmpty()) {
+                TResultBatch resultBatch = resultBatchQueue.poll();
+                futures.add(executorService.submit(() -> {
+                    List<FileScanTask> scanTasks = parse(resultBatch);
+                    return fileScanTaskQueue.addAll(scanTasks);
+                }));
+            }
+        }
+
+        if (context.getState().isError()) {
+            String collectErrorMsg = "Failed to execute metadata collection job. ";
+            String errMsgToClient = job.getMetadataJobCoord().getExecStatus().getErrorMsg();
+            this.metadataCollectionException = new StarRocksConnectorException(collectErrorMsg + errMsgToClient);
+            fileScanTaskQueue.clear();
+            return;
+        }
+
+        try {
+            for (Future<Boolean> future : futures) {
+                future.get();
+            }
+        } catch (Exception e) {
+            String parserErrorMsg = "Failed to parse iceberg file scan task. ";
+            LOG.error(parserErrorMsg, e);
+            this.metadataCollectionException = new StarRocksConnectorException(parserErrorMsg + e.getMessage());
+            fileScanTaskQueue.clear();
+        } finally {
+            while (true) {
+                if (fileScanTaskQueue.isEmpty()) {
+                    parserFinished.set(true);
+                    break;
+                }
+
+                try {
+                    Thread.sleep(50);
```

Comment:
why need to sleep?

Reply:
I think it's not a good idea that using `while(true)` to wait for the end. So I take a short sleep.